include/deal.II/numerics/smoothness_estimator.h

```diff
@@ -0,0 +1,268 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 1998 - 2018 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_smoothness_estimator_h
+#define dealii_smoothness_estimator_h
+
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/fe/fe_series.h>
+
+#include <deal.II/lac/vector.h>
+
+#include <deal.II/numerics/vector_tools.h>
+
+#include <vector>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+
+/**
+ * A namespace for various smoothness estimation strategies for hp-adaptive FEM.
+ */
+namespace SmoothnessEstimator
+{
+  /**
+   * Estimate smoothness from decay of Legendre absolute values of coefficients
+   * on the reference cell.
+   *
+   * In one dimension, the finite element solution on the reference element with
+   * polynomial degree $p$ can be written as
+   * @f[
+   *    u(x) = \sum_{j=0}^{p} a_j P_j(x)
+   * @f]
+   * where $\{P_j(x)\}$ are Legendre polynomials. The decay of the coefficients
+   * is estimated by performing the least squares fit of
+   * @f[
+   *   |a_j| \sim C exp(-\sigma j)
+   * @f]
+   * for $j=1,..,p$. The rate of the decay can be used to estimate the
+   * smoothness. For example, p-refinement will be performed if $\sigma>1$.
+   *
+   * Extension to higher dimension is done by performing the fit in each
+   * direction separately and then taking the lowest value of $\sigma$.
+   *
+   * The returned value for each vector and for each cell is $\exp(-\sigma)$.
+   *
+   * @param fe_series FESeries::Legendre object to calculate coefficients. This
+   * object needs to be initialized to have at least $p+1$ coefficients in each
+   * direction, where $p$ is the maximum polynomial degree to be used.
+   * @param dof_hander an hp::DoFHandler
```

Comment:
```suggestion
   * @param dof_hander An hp::DoFHandler
```
This just looks better when run through doxygen